runtime/lib/backend/openqasm/OpenQasmBuilder.hpp

```diff
@@ -0,0 +1,409 @@
+            // name[slice_0], ..., name[slice_n]
+            RT_ASSERT(isValidSlice(slice));
+            auto iter = slice.begin();
+            for (; iter != slice.end() - 1; iter++) {
+                oss << this->name << "[" << *iter << "], ";
+            }
+            oss << this->name << "[" << *iter << "]";
+            return oss.str();
+        }
+        default:
+            RT_FAIL("Unsupported OpenQasm register mode");
+        }
+    }
+};
+
+/**
+ * The OpenQasm gate type.
+ *
+ * @param name The name of the gate to apply from the list of supported gates
+ * (`rt_qasm_gate_map`)
+ * @param params_val Optional list of parameter values for parametric gates
+ * @param params_str Optional list of parameter names for parametric gates
+ * @param wires Wires to apply gate to
+ * @param inverse Indicates whether to use inverse of gate
+ *
+ * @note Parametric gates are currently supported via either their values or names
+ * but not both.
+ */
+class QasmGate {
+  private:
+    const std::string name;
+    const std::vector<double> params_val;
+    const std::vector<std::string> params_str;
+    const std::vector<size_t> wires;
+    const bool inverse;
+
+  public:
+    explicit QasmGate(const std::string &_name, const std::vector<double> &_params_val,
+                      const std::vector<std::string> &_params_str,
+                      const std::vector<size_t> &_wires, [[maybe_unused]] bool _inverse)
+        : name(lookup_qasm_gate_name(_name)), params_val(_params_val), params_str(_params_str),
+          wires(_wires), inverse(_inverse)
+    {
+        RT_FAIL_IF(!(params_str.empty() || params_val.empty()),
+                   "Parametric gates are currently supported via either their values or names but "
+                   "not both.");
+    }
+    ~QasmGate() = default;
+
+    [[nodiscard]] auto getName() const -> std::string { return this->name; }
+    [[nodiscard]] auto getParams() const -> std::vector<double> { return this->params_val; }
+    [[nodiscard]] auto getParamsStr() const -> std::vector<std::string> { return this->params_str; }
+    [[nodiscard]] auto getWires() const -> std::vector<size_t> { return this->wires; }
+    [[nodiscard]] auto getInverse() const -> bool { return this->inverse; }
+
+    auto toOpenQasm(const QasmRegister &qregister, [[maybe_unused]] size_t precision = 5,
+                    [[maybe_unused]] const std::string &version = "3.0") const -> std::string
+    {
+        // name(param_1, ..., param_n) qubit_1, ..., qubit_m
+        std::ostringstream oss;
+        oss << this->name;
+        if (!this->params_val.empty()) {
+            oss << "(";
+            auto iter = this->params_val.begin();
+            for (; iter != this->params_val.end() - 1; iter++) {
+                oss << std::setprecision(precision) << *iter << ", ";
+            }
+            oss << std::setprecision(precision) << *iter << ") ";
+        }
+        else if (!this->params_str.empty()) {
```

Comment:
Following my above question: why not both?

Reply:
Let's continue the discussion in the thread above :)